### PR TITLE
Implement precompile mock contracts and foundry mock util test contracts

### PR DESCRIPTION
**Pull request by mshakeg**
#### Description

Replaces https://github.com/hashgraph/hedera-smart-contracts/pull/236 for a cleaner review.

- Add foundry test support files
- Add husky commit-msg to help devs not forget signing off
- Add foundry testing readme file
- Add npm package script for easy runs of foundry, hardhat and localnode
- Add HTS mock and test contracts and use for reference foundry tests
- Add github CI action to run foundry tests

### CODE REVIEW

**Comment by mshakeg** [comment]
Hey @Nana-EC I opened this PR as the git rebase to sign-off all commits wasn't clean and ended up showing a bunch of other changes.

**Comment by Nana-EC** [comment]
> Hey @Nana-EC I opened this PR as the git rebase to sign-off all commits wasn't clean and ended up showing a bunch of other changes.
> 
> 
> 
> There are still quite a lot of TODOs documented in the code, when I free up a bit I'll create issues for those and address them.

Np.
Once you break out the issues we can see how to get this in so maybe you don't have to do it all.
Thanks

**Comment by mshakeg** [comment]
@Nana-EC do you want to merge this PR as is or should I first remove the TODOs scattered throughout the code?

File: README.md
Comment (Nana-EC) [inline README.md:1]: Revert these README changes.
You accidentally removed the valid test.

I'd advise you move this to a `FOUNDRY_TESTING.MD` file and then add a section in this README uner `## Development guidelines` pointing to it.

File: .github/workflows/test.yml
Comment (Nana-EC) [inline .github/workflows/test.yml:1]:
```suggestion
name: Foundry Tests
```

File: .github/workflows/test-workflow.yml
Comment (Nana-EC) [inline .github/workflows/test-workflow.yml:43]: I'd prefer this broken out as if there's ever a failure in Hardhat configs or local node the Foundry part wouldn't get run and if Foundry failed it would make it look like previous tests failed.
Let's just use the new workflow you added and think of how to combine them later.

File: .github/workflows/test.yml
Comment (Nana-EC) [inline .github/workflows/test.yml:3]: You can add this to run on PRs so we get the feedback.
When we turn to focus on official support we'll add more scenarios to ensure coverage on release and push also
```suggestion
on: 
  pull_request:
    branches: [main]
  workflow_dispatch
```

File: .github/workflows/test.yml
Comment (Nana-EC) [reply .github/workflows/test.yml:1]: Also rename file to `foundry_test.yml`

File: test/foundry/mocks/libraries/HederaTokenValidation.sol
Comment (Nana-EC) [inline test/foundry/mocks/libraries/HederaTokenValidation.sol:255]: Just open a ticket to capture this. You can make code comment the ticket

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol
Comment (Nana-EC) [inline test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:121]: Just open a ticket to capture this. You can make code comment the ticket

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Some minor docs, naming and workflow changes to make it ready to get in

**Comment by Nana-EC** [comment]
> @Nana-EC do you want to merge this PR as is or should I first remove the TODOs scattered throughout the code?

@mshakeg gonna work with you to get this in hopefuly this week.
I've reviewed and I think getting DCO (remember to [sign-off](https://github.com/hashgraph/.github/blob/main/CONTRIBUTING.md#sign-off)) to pass and addressing these minor comments and we can get this in

File: .github/workflows/test-workflow.yml
Comment (mshakeg) [reply .github/workflows/test-workflow.yml:43]: Ok, it's currently setup such that `forge compile` will attempt to compile `./contracts` which depend on dependencies installed in the `node_modules`, so I'll move this to execute once after a given job.

Comment (mshakeg) [reply .github/workflows/test-workflow.yml:43]: I removed this, as `foundry_test.yml` suffices

File: README.md
Comment (mshakeg) [reply README.md:1]: Which "valid test" did I accidentially remove?

File: test/foundry/mocks/libraries/HederaTokenValidation.sol
Comment (mshakeg) [reply test/foundry/mocks/libraries/HederaTokenValidation.sol:255]: I opened a few issues to capture all the TODOs, but I think it would be best to leave the TODO comments for the developer working on the issue to more easily understand where the issue is.

File: README.md
Comment (Nana-EC) [reply README.md]: Mistyped test, meant README section content

File: test/foundry/mocks/libraries/HederaTokenValidation.sol
Comment (Nana-EC) [reply test/foundry/mocks/libraries/HederaTokenValidation.sol:255]: That's fine, no biggie

**Comment by Nana-EC** [comment]
> @Nana-EC thanks for the review, I added a husky hook to prevent committing without signing off as I don't always remember to do so.

@mshakeg Looks good to go. Just waiting on CI to pass and I'll merge this in.
Appreciate the contributions here and as always if you see any gaps feel free to open tickets to get on our radar or awesome PRs like this to accelerate  the process.
Thanks

**Comment by Nana-EC** [comment]
> > @Nana-EC thanks for the review, I added a husky hook to prevent committing without signing off as I don't always remember to do so.
> 
> @mshakeg Looks good to go. Just waiting on CI to pass and I'll merge this in. Appreciate the contributions here and as always if you see any gaps feel free to open tickets to get on our radar or awesome PRs like this to accelerate the process. Thanks

And just after I commented of course all the CI tests fail - https://github.com/hashgraph/hedera-smart-contracts/actions/runs/6546507450/job/17787629512?pr=462

**Comment by mshakeg** [comment]
@Nana-EC thanks, will do. Once this is in, I'll continue with the `foundry-example` project in the json rpc relay

File: .gitignore
Comment (AlfredoG87) [inline .gitignore:33]: could we add a blank line at the end of the file?

File: foundry.toml
Comment (AlfredoG87) [inline foundry.toml:12]: blank line at EOF missing

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol
Comment (AlfredoG87) [inline test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:96]: Why this commented code? is it intentional?

**Review by AlfredoG87 — COMMENTED** [review]
Incredible work, I will have to come back later to complete review, but overall it looks really good. 
Only 2 nits and 1 question.

File: test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol
Comment (mshakeg) [reply test/foundry/mocks/hts-precompile/HederaNonFungibleToken.sol:96]: It can be removed the validation is/should be done in the `HtsSystemContractMock` contract

**Review by Nana-EC — APPROVED** [review]
LG
Nice work

**Comment by Nana-EC** [comment]
> @Nana-EC thanks, will do. Once this is in, I'll continue with the `foundry-example` project in the json rpc relay

@mshakeg you're all set, only took a little bit 🤣

**Pull request merged**
